optuna/integration/botorch.py

```diff
@@ -0,0 +1,468 @@
+
+        self._candidates_func = candidates_func
+        self._constraints_func = constraints_func
+        self._independent_sampler = independent_sampler or RandomMultiObjectiveSampler()
+        self._n_startup_trials = n_startup_trials
+
+        self._trial_constraints: Dict[int, Tuple[float, ...]] = {}
+        self._study_id: Optional[int] = None
+        self._search_space = IntersectionSearchSpace()
+
+    def infer_relative_search_space(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+    ) -> Dict[str, BaseDistribution]:
+        if self._study_id is None:
+            self._study_id = study._study_id
+        if self._study_id != study._study_id:
+            # Note that the check below is meaningless when `InMemortyStorage` is used
+            # because `InMemortyStorage.create_new_study` always returns the same study ID.
+            raise RuntimeError("BoTorchSampler cannot handle multiple studies.")
+
+        return self._search_space.calculate(study, ordered_dict=True)  # type: ignore
+
+    def _update_trial_constraints(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trials: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> None:
+        # Since trial constraints are computed on each worker, constraints should be computed
+        # deterministically.
+
+        assert self._constraints_func is not None
+
+        for trial in trials:
+            number = trial.number
+            if number not in self._trial_constraints:
+                constraints = self._constraints_func(study, trial)
+
+                if not isinstance(constraints, (tuple, list)):
+                    raise TypeError("Constraints must be a tuple or list.")
+
+                constraints = tuple(constraints)
+                self._trial_constraints[number] = constraints
+
+    def sample_relative(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        trial: "multi_objective.trial.FrozenMultiObjectiveTrial",
+        search_space: Dict[str, BaseDistribution],
+    ) -> Dict[str, Any]:
+        assert isinstance(search_space, OrderedDict)
+
+        trials = [t for t in study.get_trials(deepcopy=False) if t.state == TrialState.COMPLETE]
+        if self._constraints_func is not None:
+            self._update_trial_constraints(study, trials)
+
+        if len(search_space) == 0:
+            return {}
+
+        n_trials = len(trials)
+        if n_trials < self._n_startup_trials:
+            return {}
+
+        trans = _SearchSpaceTransform(search_space)
+
+        values = numpy.empty((n_trials, study.n_objectives), dtype=numpy.float64)
+        params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
+        if self._constraints_func is not None:
+            n_objective_constraints = len(next(iter(self._trial_constraints.values())))
```

Comment:
I preferred `n_constraints` to `n_objective_constraints` since it is shorter and the constraints are also for the input.

Reply:
Ah, thanks, fixed.